setup-local/src/binaryControl.js

```diff
@@ -0,0 +1,232 @@
+  }
+
+  /**
+   * Generates the args to be provided for the Local Binary based on the operation, i.e.
+   * start/stop.
+   * These are generated based on the input state provided to the Binary Control.
+   */
+  _generateArgsForBinary() {
+    const {
+      accessKey: key,
+      localArgs,
+      localIdentifier,
+      localLoggingLevel: verbose,
+      localTesting: binaryAction,
+    } = this.stateForBinary;
+
+    let argsString = `--key ${key} --only-automate --ci-plugin GitHubAction `;
+
+    switch (binaryAction) {
+      case LOCAL_TESTING.START: {
+        if (localArgs) argsString += `${localArgs} `;
+        if (localIdentifier) argsString += `--local-identifier ${localIdentifier} `;
+        if (verbose) {
+          this._generateLogFileMetadata();
+          argsString += `--verbose ${verbose} --log-file ${this.logFilePath} `;
+        }
+        break;
+      }
+      case LOCAL_TESTING.STOP: {
+        if (localIdentifier) argsString += `--local-identifier ${localIdentifier} `;
+        break;
+      }
+      default: {
+        throw Error('Invalid Binary Action');
+      }
+    }
+
+    this.binaryArgs = argsString;
+  }
+
+  /**
+   * Triggers the Local Binary. It is used for starting/stopping.
+   * @param {String} operation start/stop operation
+   */
+  async _triggerBinary(operation) {
+    let triggerOutput = '';
+    let triggerError = '';
+    await exec.exec(
+      `${LOCAL_BINARY_NAME} ${this.binaryArgs} --daemon ${operation}`,
+      [],
+      {
+        listeners: {
+          stdout: (data) => {
+            triggerOutput += data.toString();
+          },
+          stderr: (data) => {
+            triggerError += data.toString();
+          },
+        },
+      },
+    );
+
+    return {
+      output: triggerOutput,
+      error: triggerError,
+    };
+  }
+
+  /**
+   * Downloads the Local Binary, extracts it and adds it in the PATH variable
+   */
+  async downloadBinary() {
+    try {
+      if (Utils.checkToolInCache(LOCAL_BINARY_NAME)) {
+        core.info('BrowserStackLocal binary already exists in cache. Using that instead of downloading again...');
+        return;
+      }
+      await this._makeDirectory();
+      core.info('Downloading BrowserStackLocal binary...');
+      const downloadPath = await tc.downloadTool(this.binaryLink, path.resolve(this.binaryFolder, 'binaryZip'));
+      const extractedPath = await tc.extractZip(downloadPath, this.binaryFolder);
+      core.info(`BrowserStackLocal binary downloaded & extracted successfuly at: ${extractedPath}`);
+      const cachedPath = await tc.cacheDir(extractedPath, LOCAL_BINARY_NAME, '1.0.0');
+      core.addPath(cachedPath);
+      this.binaryPath = extractedPath;
+    } catch (e) {
+      throw Error(`BrowserStackLocal binary could not be downloaded due to ${e.message}`);
+    }
+  }
+
+  /**
+   * Starts Local Binary using the args generated for this action
+   */
+  async startBinary() {
+    try {
+      this._generateArgsForBinary();
```

Comment:
is it really required to have most of the code in try-catch? might be missing context but should we not go for specific exceptions instead of generic ones?

Reply:
refactoring that.

Reply:
The reason why it's enclosed in the try-catch block is that it:
1. begins with the generation of `args` for binary. And that doesn't have its own `error handling`. This is a main operation which actually starts (or stops - the other function below it)... and they throw errors at different stages. 
2. Its possible that `generation of args` failed or `triggering of binary resulted in an error`... or `parsing the response failed`. All are necessary steps to be checked for error condition.

Reply:
3. The methods starting with `_` are trying to act like private methods. They don't have their own handling of errors. The parent method, or the `public` method using them should have the handling.

Reply:
resolving this.